File: 2568-minimum-fuel-cost-to-report-to-the-capital/2568-minimum-fuel-cost-to-report-to-the-capital.py
import math
from typing import List

class Solution:
    def minimumFuelCost(self, roads: List[List[int]], seats: int) -> int:
        n = len(roads) + 1
        adj = [[] for _ in range(n)]

        for u, v in roads:
            adj[u].append(v)
            adj[v].append(u)

        total_fuel = 0

        def dfs(node: int, parent: int) -> int:
            nonlocal total_fuel
            people = 1

            for child in adj[node]:
                if child != parent:
                    people += dfs(child, node)

            if node != 0:
                total_fuel += math.ceil(people / seats)

            return people

        dfs(0, -1)
        return total_fuel

File: 2568-minimum-fuel-cost-to-report-to-the-capital/2568-minimum-fuel-cost-to-report-to-the-capital.java
import java.util.*;

class Solution {
    private long totalFuel = 0;
    
    public long minimumFuelCost(int[][] roads, int seats) {
        int n = roads.length + 1;
        List<List<Integer>> adj = new ArrayList<>();
        
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        
        for (int[] road : roads) {
            int u = road[0], v = road[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        
        dfs(0, -1, adj, seats);
        return totalFuel;
    }
    
    private int dfs(int node, int parent, List<List<Integer>> adj, int seats) {
        int people = 1;
        
        for (int child : adj.get(node)) {
            if (child != parent) {
                people += dfs(child, node, adj, seats);
            }
        }
        
        if (node != 0) {
            // totalFuel += (people + seats - 1) / seats;
            totalFuel += (long) Math.ceil((double) people / seats);
        }
        
        return people;
    }
}